### PR TITLE
DOC: re-add api example

### PR DESCRIPTION
**Pull request by jklymak**
## PR summary

This re-introduces the example in `api/index`, and adds a pyplot counterexample.  

As discussed in #26623 - this provides concrete examples of the abstract concepts being discussed and makes the API pages more self-contained.  

## PR checklist

- [ ] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

**Comment by jklymak** [comment]
> I also think that since this is a doc regression it should have two approvals.

The example originally came in https://github.com/matplotlib/matplotlib/pull/19727.  It was removed in #26402 without mention of the removal in the PR description, or discussion on the strength of one review.  I'm not clear how re-instating it is considered a regression.

**Comment by timhoffm** [comment]
> The example originally came in https://github.com/matplotlib/matplotlib/pull/19727. It was removed in https://github.com/matplotlib/matplotlib/pull/26402 without mention of the removal in the PR description, or discussion

In https://github.com/matplotlib/matplotlib/pull/26402, motivated the change:

> Reduce the interface description to very compact two-sentence descriptions and refer to the interface page for everything else.
> 
> The focus here is to point the users to relevant API resources for the respective interfaces, not to discuss these interfaces in detail. The latter is the task of https://matplotlib.org/devdocs/users/explain/figure/api_interfaces.html

This page is the entry point for the *API reference*, the collection of all function / class docstrings. Categorizing according to https://documentation.divio.com/ it is information-oriented, not learning-oriented or explanation-oriented.

My train of thought for deciding on the contents of the page is:

1. In the very basic form, the **Modules** section would be enough.
2. However, the user-facing interfaces are what people will need most of the time. Prominently highlight direct the user to the most relevant content:
   
   grafik

3. For people who already know the interfaces, this would be enough. But we know that a lot of users have no clear idea on this. To help them decide between the worlds, we sould give a very rough idea and link to the detailed description if they want to /need to learn more. - But this additional interface description should be as terse as possible because it distracts from the actual relevant links in 2. and 1.


In particular, I believe a visual plot would be harming the cause here. It draws a lot of visual attention and does not explain anything about the API or Interfaces. I would be +/-0 on adding minimal pure code-blocks to the existing interface cards

```
fig, ax = plt.subplots()
ax.plot(x, y)
ax.set_title('Sample plot')
plt.show()
```

```
plt.plot(x, y)
plt.title('Sample plot')
plt.show()
```

File: doc/api/index.rst [hunk 1]
```diff
@@ -24,21 +24,56 @@ use cases.
...
+      y = np.sin(x*np.pi)
+      # Create a Figure and an Axes:
+      fig, ax = plt.subplots(figsize=(3,2), layout='constrained')
+      # Use the Axes to plot and label:
+      ax.plot(x, y)
+      ax.set_xlabel('t [s]')
+      ax.set_ylabel('S [V]')
+      ax.set_title('Sine wave')
+      # Change a property of the Figure:
+      fig.set_facecolor('lightsteelblue')
```
[the remark this answers is not in the document]
Comment (jklymak) [reply doc/api/index.rst:56]: Two column is fine, though it looks pretty cramped. 

Re `ax.set` that seems an advanced usage to me, whereas I thought we wanted this to be straightforward?  ax.,set also does not allow one to customize anything in the set methods (eg fontsize in title etc), so I'm not sure it should become canonical usage.

**Comment by timhoffm** [comment]
> Information doesn't mean anything without some context.

Fully agree and that's why I added bullet point 3. in my reasoning. The key here is *some*. Can we agree on the principle that *some* means "enough to help understanding the main information, but as little as possible to not distract from the main information"? - The judgement call, and where I think we disargee is, is what is "enough".

> I strongly feel the example helps set the context for what is being discussed on this page. Our API is very bizarre in that 95% of it is on the Axes objects we create, and showing a short example of that on a top-level landing page helps orient the reader.

For the context, the priority, determined by relevance, is:

1. **Short fundamental description**: we agree on this
2. **Code snippet to illustrate how the API looks**: as written above, I'm +/-0 on this - I don't think it's essential, but it's compact enough to not distract.
3. **Resulting plot of the code example**: The visual plot does not really help in understanding the API. I propose that
   ```
   plt.plot(x, y)
   plt.title('Sample plot')
   plt.show()
   ```
   is self-explanatory enough. An added problem of the visual is that you have to provide fully working code, which means you have to clutter the code with import and data generation, which clearly do not add to understanding the API.
   IMHO this is too much clutter for almost no gain.

**Comment by jklymak** [comment]
I think it is a strange aesthetic to just show the code, but not show the code result in a plotting library. Code tends to be a lot easier to parse if you can see its result.   You say it's "cluttered", and "distracts".  But from what?  There is no other content on this page - the rest of this page is just an index (that is repeated verbatim on the left sidebar).

**Comment by timhoffm** [comment]
> I think it is a strange aesthetic to just show the code, but not show the code result in a plotting library.

It depends on what you want to achieve with that code. There are lots of **Examples** sections in our docstrings that do not generate the plot.

File: doc/api/index.rst [hunk 2]
```diff
@@ -24,21 +24,56 @@ use cases.
...
+
+      x = np.arange(0, 4, 0.05)
+      y = np.sin(x*np.pi)
+      # Create a Figure and an Axes:
+      fig, ax = plt.subplots(figsize=(3,2), layout='constrained')
+      # Use the Axes to plot and label:
+      ax.plot(x, y)
+      ax.set_xlabel('t [s]')
+      ax.set_ylabel('S [V]')
+      ax.set_title('Sine wave')
```
Comment (tacaswell) [inline doc/api/index.rst:54]:
```suggestion
      ax.set_title('Sine wave (explicit)')
```

File: doc/api/index.rst [hunk 3]
```diff
@@ -24,21 +24,56 @@ use cases.
...
+
+      x = np.arange(0, 4, 0.05)
+      y = np.sin(x*np.pi)
+      # Create a Figure of a given size:
+      plt.figure(figsize=(3, 2), layout='constrained')
+      # plot on a default Axes on the Figure:
+      plt.plot(x, y)
+      plt.xlabel('t [s]')
+      plt.ylabel('S [V]')
+      plt.title('Sine wave')
```
Comment (tacaswell) [inline doc/api/index.rst:73]:
```suggestion
      plt.title('Sine wave (implicit)')
```

File: doc/api/index.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (tacaswell) [reply doc/api/index.rst:56]: There is a lot of value in the 1:1 mapping between the lines so am very 👎🏻 on changing to use `ax.set`.

At least for me clicking between the tabs and seeing what does/does not change is way easier to see the differences than trying to scan between two columns.

**Comment by jklymak** [comment]
The point of this PR was to re-add the figure.  #26671 did not re-add the figure, and obviously there is some disagreement about whether figures are desirable in our docs.

**Comment by tacaswell** [comment]
I think the longer example with an image is actually quite helpful here, particularly as written with the 1:1 mapping of the lines of code.  By flipping back and forth between the tabs it is easy to see what the required changes are and than you get the same output either way (the only change I would suggest would be to make the titles different so it is clear that the image _is_ changing between the tabs, currently the only way to tell is to mouse over them and see the urls are different).

We (the developers) know that the APIs are interchangeable, but that may not be clear to all of our users so showing that you can get exactly the same output using either path is a good thing.  It is also good to show that the explicit API is not also necessarily more verbose. 


----

We probably should re-write some of the examples in https://matplotlib.org/devdocs/users/explain/figure/api_interfaces.html using this 2 tab style.

**Comment by tacaswell** [comment]
tabs

At least to my eye the tabs really highlight exactly what is changing between the two and widely spaced (or vertically stacked) code blocks are much much harder to compare.

**Comment by jklymak** [comment]
At the end of the day, these are all taste issues.  I think the final documentation say is @tacaswell, so he should just state his preference.  

I will point out some usability issues with the current links - I think that `Figure` should be changed back to `matplotlib.figure` because the first links to the init docs https://matplotlib.org/stable/api/figure_api.html#matplotlib.figure.Figure which no one ever uses, and the latter links to the top of the API page, which has the TOC for Figure methods that people do us.  Same for `Axes`/`matplotlib.axes`.

**Comment by jklymak** [comment]
Again, whether something looks better tabbed or side-by-side is not subject to any hard and fast rules.  FWIW I prefer side-by-side, but I wouldn't claim that my preference is "right" or a "best-practice".  I had a slight preference for tabbed here just because it was more compact.

> We have other pages in the docs where we show all the things, how does doing that here help to more effectively communicate the differences between the two APIs?

The original impetus to improving this page (https://github.com/matplotlib/matplotlib/pull/19727), and adding an example, is that this is the top-level API landing page, arrived at via "Reference".  Lots of people who think they are familiar with Matplotlib probably land on "Reference", and then don't know where to go.  I distinctly remember getting started with our API, and not understanding that most of it was in the `Axes` class.  So while I _knew_ that there was such a thing as `ax.plot` I did not know where to find its documentation, other than `pyplot.plot`, which is not quite the same thing. 

In my mind, the goal of these improvements was _not_ primarily to "communicate the difference between the two APIs".  Indeed the original version of this had all the different API discussion _below_ the TOC (https://matplotlib.org/3.5.3/api/index.html).  Rather the goal was to provide a quick guide/reminder of our API using code snippets, and their results, that user may have seen elsewhere, be that in our docs, stack overflow etc.  I agree that the difference between the APIs needs to be communicated, but as a small part of the larger goal of providing a guide the API reference.

**Comment by jklymak** [comment]
The data creation is pretty trivial - I hope we can assume our readers have basic familiarity with numpy.  Further, it makes it clear what kind of objects x and y are, versus completely unspecified. 

As for the setters, I think labelling a plot is best practice and common enough to warrant a couple of lines here.  

As for tabs versus side by side, in addition to size issues, we are trying to discourage one of the APIs - recall the original version didn't even have the pyplot API example.  In fact for the purpose of indexing I'm not even sure it's needed at all since the pyplot API is easy to find 

I wouldn't object to re organizing the API on the page, but that is somewhat orthogonal here

**Comment by jklymak** [comment]
Again, we disagree fundamentally on most of these points, none of which have an objective criteria.   Someone, presumably @tacaswell, will have to decide what they want this page to look like.

**Comment by jklymak** [comment]
Sure, if that falls under that bailiwick, I'll close this.

**Pull request closed without merging**